File: bindings/c/src/event.cpp
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

#include <nux/event.h>
#include "handles.hpp"

NUXC_EXTERNC
nux_event_t*
nux_event_create (
    nux_event_type_t initial_event_type
) {
    return new nux_event_t{nux::event{nux_impl_map(initial_event_type)}};
}

NUXC_EXTERNC
void
nux_event_destroy (
    nux_event_t* event
) {
    delete event;
}

NUXC_EXTERNC
nux_event_type_t
nux_event_get_event_type (
    nux_event_t* event
) {
    return nux_impl_map(event->handle.get_event_type());
}

File: bindings/rs/nux/src/event.rs
// Copyright (c) 2025 Daniel T. McGinnis
// SPDX-License-Identifier: MIT

use crate::bindings::*;

pub enum EventType {
    KeyEvent,
    CloseEvent
}

pub(crate) fn map_event_type_t (
    event_type: nux_event_type_t
) -> EventType {
    return match event_type {
        nux_event_type_t_NUXC_EVENT_TYPE_KEY_EVENT => EventType::KeyEvent,
        nux_event_type_t_NUXC_EVENT_TYPE_CLOSE_EVENT => EventType::CloseEvent,
        _ => unreachable!()
    };
}

pub(crate) fn map_event_type (
    event_type: EventType
) -> nux_event_type_t {
    return match event_type {
        EventType::KeyEvent => nux_event_type_t_NUXC_EVENT_TYPE_KEY_EVENT,
        EventType::CloseEvent => nux_event_type_t_NUXC_EVENT_TYPE_CLOSE_EVENT,
        _ => unreachable!()
    };
}

pub(crate) trait AccessEventType {
    fn get_event_type (
        &self
    ) -> EventType;
}

pub struct Event {
    native_handle: *mut nux_event_t
}

impl Event {
    pub fn new (
        event_type: EventType
    ) -> Self {
        return Event {
            native_handle: unsafe {
                nux_event_create(map_event_type(event_type))
            }
        };
    }
}

impl Drop for Event {
    fn drop (
        &mut self
    ) {
        unsafe {
            nux_event_destroy(self.native_handle)
        };
    }
}

impl AccessEventType for Event {
    fn get_event_type (
        &self
    ) -> EventType {
        return map_event_type_t(
            unsafe { nux_event_get_event_type(self.native_handle) }
        );
    }
}
